adapters/kobler/kobler.go

```diff
@@ -0,0 +1,146 @@
+package kobler
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint    string
+	devEndpoint string
+}
+
+const (
+	DEV_BIDDER_ENDPOINT = "https://bid-service.dev.essrtb.com/bid/prebid_server_rtb_call"
+	SUPPORTED_CURRENCY  = "USD"
+)
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint:    config.Endpoint,
+		devEndpoint: DEV_BIDDER_ENDPOINT,
+	}
+
+	return bidder, nil
+}
+
+func (a adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var requestData []*adapters.RequestData
+	var errors []error
+
+	if !contains(request.Cur, SUPPORTED_CURRENCY) {
+		request.Cur = append(request.Cur, SUPPORTED_CURRENCY)
+	}
+
+	for i := range request.Imp {
+		if err := convertImpCurrency(&request.Imp[i], reqInfo); err != nil {
+			errors = append(errors, err)
+			return nil, errors
+		}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+
+	requestData = append(requestData, &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.getEndpoint(request),
+		Body:    requestJSON,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+		Headers: headers,
+	})
+
+	return requestData, nil
+}
+
+func (a adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent || responseData.Body == nil {
+		return nil, nil
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: getMediaTypeForBid(bid),
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func (a adapter) getEndpoint(request *openrtb2.BidRequest) string {
+	if request.Test == 1 {
+		return a.devEndpoint
+	}
```

Comment:
I haven't seen this behavior before. Why do you require a separate endpoint for test requests?

Reply:
We have a test campaign in our DEV environment that's set to bid on a specific article. For this to work, we have to do some manual manipulation of the article in our system and this is easier to do in the DEV environment.

This is the same as for our [Prebid.js adapter](https://docs.prebid.org/dev-docs/bidders/kobler.html).

Reply:
IMO it's good to add some comments:
// Use a separate endpoint for testing purposes in the DEV environment.
// Required due to Kobler's internal test campaign setup.
because: Introducing a separate endpoint (devEndpoint) for test requests is unusual in Prebid Server, since standard tests use the same endpoint as the production environment.

Reply:
@przemkaczmarek I have added this now, thank you!